Add spin_and_wait_for_matched to PublicationManager and update test c…

Barry-Xu-2018 opened the pull request:
Address #790

MichaelOrlov commented:
@Barry-Xu-2018 I am kind of confused about your implementation and this is not what I expected to see.
Why do you need to spin node for PublicationManager?
It does make sense to spin node only for subscription.

emersonknapp commented on rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:
```diff
@@ -76,26 +79,78 @@ class PublicationManager
   std::shared_ptr<rclcpp::Node> pub_node_;
   using PublicationF = std::function<void (bool)>;
-  std::vector<PublicationF> publishers_;
+  std::vector<PublicationF> publishers_fcn_;
```
Is there any reason why we should maintain two separate vectors? I would prefer 

```
std::vector<std::pair<rclcpp::PublisherBase::SharedPtr, PublicationF>> publishers_;
```

This way it's impossible for the vectors to get out of sync.

emersonknapp commented on rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:
```diff
@@ -76,26 +79,78 @@ class PublicationManager
       pub_fcn(verbose);
     }
+    rclcpp::executors::SingleThreadedExecutor exec;
+    exec.spin_node_some(pub_node_);
```
Is spinning necessary for publishing?

Barry-Xu-2018 replied:
Not necessary.

emersonknapp commented on rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:
```diff
@@ -76,26 +79,78 @@ class PublicationManager
     }
+    rclcpp::executors::SingleThreadedExecutor exec;
+    exec.spin_node_some(pub_node_);
+    std::this_thread::sleep_for(100ms);
```
Is 100ms meaningful? Where does this number come from? Why do we need to sleep? Does this make the publishing "bursty"?

Barry-Xu-2018 replied:
Not necessary. I forgot to remove them.

Barry-Xu-2018 commented:
@MichaelOrlov 

> I am kind of confused about your implementation and this is not what I expected to see.
Why do you need to spin node for PublicationManager?
It does make sense to spin node only for subscription.

According to your suggestion in https://github.com/ros2/rosbag2/issues/790#issuecomment-868945959
```
For the Problem 1 related to the discovery it would be more appropriate to add wait_for_matched(..) API for the PublicationManager. Similar as made spin_and_wait_for_matched in SubscriptionManager.
```
My understanding is to add similar API like spin_and_wait_for_matched in SubscriptionManager to PublicationManager.  
Spin node is to confirm the number of connected subscription (from rosbag) for specified publisher.

MichaelOrlov commented:
@Barry-Xu-2018 It seems you misinterpreted my suggestion. I intentionally mentioned `wait_for_matched(..)` API without word spin. Also mentioned **`similar`** as `spin_and_wait_for_matched(..)`. I meant that `similar` doesn't means "the same as".

Anyway, thank you for your PR, I will take a look more closely on it since you have addressed some review comments.

Barry-Xu-2018 commented:
@MichaelOrlov 

> It seems you misinterpreted my suggestion. I intentionally mentioned wait_for_matched(..) API without word spin. Also mentioned similar as spin_and_wait_for_matched(..). I meant that similar doesn't means "the same as".

Sorry for my misunderstanding.   
Spin is  unnecessary.

> Anyway, thank you for your PR, I will take a look more closely on it since you have addressed some review comments.

Please wait a moment for the review.   
I will update codes to remove spin.

Barry-Xu-2018 commented:
@MichaelOrlov @emersonknapp 

Addressed your comments.   
Please review again.

MichaelOrlov commented on rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:
```diff
@@ -71,31 +72,90 @@ class PublicationManager
+    using clock = std::chrono::steady_clock;
+    auto start = clock::now();
+
+    rclcpp::executors::SingleThreadedExecutor exec;
```
It seems forgot to remove 
```
rclcpp::executors::SingleThreadedExecutor exec;
```

Barry-Xu-2018 replied:
Sorry. My mistake.

MichaelOrlov commented on rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:
```diff
@@ -71,31 +72,90 @@ class PublicationManager
+      return true;
+    }
+
+    auto sleep_time = std::chrono::milliseconds(100);
```
I would recommend to reduce default sleep timeout. I would chose some number from 1 to 10 mSec.

Barry-Xu-2018 replied:
Okay. I change to use 10ms as default.

MichaelOrlov reviewed (commented):
@Barry-Xu-2018 LGTM among a few nitpicks.

MichaelOrlov reviewed (approved):
@Barry-Xu-2018 All looks good now.
Thank you for your contribution.

MichaelOrlov commented on rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:
```diff
@@ -71,31 +72,89 @@ class PublicationManager
+  /// \param timeout Maximum time duration during which discovery should happen.
+  /// \param n_subscribers_to_match Number of subscribers publisher should have for match.
+  /// \return true if not find publisher by topic name or publisher has specified number of
+  ///    subscribers, otherwise false.
```
Ups. I overlooked mistake in description for return value. It should be other way around. i.e. return false if fail and true if success. 
@Barry-Xu-2018 Could you please fix it before merge?

Barry-Xu-2018 replied:
Thank you.  
The description about return value is incorrect.

MichaelOrlov commented:
Gist: https://gist.githubusercontent.com/MichaelOrlov/8f92e1829d0071ceb0844bff7ad53c52/raw/e7160639b7b34096e84da97c39d90e499def5c70/ros2.repos

BUILD args: --packages-up-to rosbag2_test_common  rosbag2_tests
TEST args: --packages-select rosbag2_test_common  rosbag2_tests
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/8641/
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14826)](http://ci.ros2.org/job/ci_linux/14826/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9576)](http://ci.ros2.org/job/ci_linux-aarch64/9576/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12492)](http://ci.ros2.org/job/ci_osx/12492/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14981)](http://ci.ros2.org/job/ci_windows/14981/)
